harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -117,15 +296,100 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
+        if self.context.distributed.size > 1:
+            with self.prof.record_timing("average_training_metrics"):
+                batch_metrics = pytorch._combine_and_average_training_metrics(
+                    self.context.distributed, training_metrics
+                )
+        else:
+            batch_metrics = training_metrics
 
-    @classmethod
-    def supports_mixed_precision(cls: Type["PyTorchTrialController"]) -> bool:
-        return True
+        metrics = det.util.make_metrics(None, batch_metrics)
+
+        # Ignore batch_metrics entirely for custom reducers; there's no guarantee that per-batch
+        # metrics are even logical for a custom reducer.
+        with self.prof.record_timing("reduce_metrics"):
+            metrics["avg_metrics"].update(
+                pytorch._convert_metrics_to_numpy(self.context.reduce_metrics(for_training=True))
+            )
+
+        if not self.is_chief:
+            return {}
+
+        # Only report on the chief worker
+        avg_metrics = metrics.get("avg_metrics", {})
+        batch_metrics = metrics.get("batch_metrics", [])
+
+        self.metric_writer.on_train_step_end(
+            self.state.batches_trained,
+            avg_metrics,
+            batch_metrics,
+        )
+        self.core_context.train.report_training_metrics(
+            steps_completed=self.state.batches_trained,
+            metrics=avg_metrics,
+            batch_metrics=batch_metrics,
+        )
+        return metrics
+
+    def _is_best_validation(self, now: float, before: Optional[float]) -> bool:
+        if before is None:
+            return True
+
+        return (now < before) if self.smaller_is_better else (now > before)
+
+    def _on_epoch_start(self, epoch_idx: int) -> None:
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_start"
+            ):
+                sig = signature(callback.on_training_epoch_start)
+                if sig.parameters:
+                    callback.on_training_epoch_start(epoch_idx)
+                else:
+                    logging.warning(
+                        "on_training_epoch_start() without parameters is deprecated"
+                        " since 0.17.8. Please add epoch_idx parameter."
+                    )
+                    callback.on_training_epoch_start()  # type: ignore[call-arg]
+
+    def _on_epoch_end(self, epoch_idx: int) -> None:
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_end"
+            ):
+                callback.on_training_epoch_end(epoch_idx)
+
+    def _checkpoint(self, already_exiting: bool) -> None:
+        if self.is_chief:
+            self.core_context.train.set_status("checkpointing")
+        self.state.last_ckpt = self.state.batches_trained
+
+        uuid = ""
```

Comment:
nit: can you put this in an `else` statement matching the `if self.is_chief` indentation?

Or at least inside the `try:` block.

But where it's at now it looks like it's going to be used in the `except` block or afterwards, which is really confusing.

Or, best yet:

```python
if self.is_chief:
    ...
    _ = self.context.distributed.broadcast(uuid)
else:
    uuid = self.context.distributed.broadcast(None)
```